heft-plugins/heft-jest-plugin/src/JestPlugin.ts

```diff
@@ -0,0 +1,422 @@
+      },
+      pathResolutionMethod: PathResolutionMethod.NodeResolve
+    };
+
+    return new ConfigurationFile<IHeftJestConfiguration>({
+      projectRelativeFilePath: 'config/jest.config.json',
+      jsonSchemaPath: schemaPath,
+      propertyInheritance: {
+        moduleNameMapper: {
+          inheritanceType: InheritanceType.custom,
+          inheritanceFunction: shallowObjectInheritanceFunc
+        },
+        transform: {
+          inheritanceType: InheritanceType.custom,
+          inheritanceFunction: shallowObjectInheritanceFunc
+        },
+        globals: {
+          inheritanceType: InheritanceType.custom,
+          inheritanceFunction: deepObjectInheritanceFunc
+        }
+      },
+      jsonPathMetadata: {
+        // string
+        '$.dependencyExtractor': nodeResolveMetadata,
+        '$.globalSetup': nodeResolveMetadata,
+        '$.globalTeardown': nodeResolveMetadata,
+        '$.moduleLoader': nodeResolveMetadata,
+        '$.snapshotResolver': nodeResolveMetadata,
+        '$.testResultsProcessor': nodeResolveMetadata,
+        '$.testRunner': nodeResolveMetadata,
+        '$.filter': nodeResolveMetadata,
+        '$.runner': nodeResolveMetadata,
+        '$.prettierPath': nodeResolveMetadata,
+        '$.resolver': nodeResolveMetadata,
+        '$.testSequencer': nodeResolveMetadata,
+        // string[]
+        '$.setupFiles.*': nodeResolveMetadata,
+        '$.setupFilesAfterEnv.*': nodeResolveMetadata,
+        '$.snapshotSerializers.*': nodeResolveMetadata,
+        // reporters: (path | [ path, options ])[]
+        '$.reporters[?(@ !== "default")]*@string()': nodeResolveMetadata, // string path, excluding "default"
+        '$.reporters.*[?(@property == 0 && @ !== "default")]': nodeResolveMetadata, // First entry in [ path, options ], excluding "default"
+        // watchPlugins: (path | [ path, options ])[]
+        '$.watchPlugins.*@string()': nodeResolveMetadata, // string path
+        '$.watchPlugins.*[?(@property == 0)]': nodeResolveMetadata, // First entry in [ path, options ]
+        // transform: { [regex]: path | [ path, options ] }
+        '$.transform.*@string()': nodeResolveMetadata, // string path
+        '$.transform.*[?(@property == 0)]': nodeResolveMetadata // First entry in [ path, options ]
+      }
+    });
+  }
+
+  public apply(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    options?: IJestPluginOptions
+  ): void {
+    // TODO: Remove when newer version of Heft consumed
+    if (options) {
+      JsonSchema.fromFile(PLUGIN_SCHEMA_PATH).validateObject(options, 'config/heft.json');
+    }
+
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.afterCompile.tapPromise(PLUGIN_NAME, async () => {
+          // Write the data file used by jest-build-transform
+          if (build.properties.isTypeScriptProject) {
```

Comment:
What if it is a TS project, but someone runs `heft test` without running `heft build`? The `_validateJestTypeScriptDataFileIfExistsAsync` won't throw an error, so will they get potentially confusing error messages from the transformer?

I was initially thinking we should update the schema of the `jest-typescript-data.json` to have a concrete indication that the project was built and that it is or isn't a TS project.

Reply:
Do errors in transformers not get bubbled up? I modified the transformer to blow up with an error if the file can't be found, asking the dev if they forgot to run `heft build`.

Reply:
Just tested this. They don't bubble up at all. Will funnel through the data file and validate before kicking jest